zipkin/src/main/java/zipkin/internal/Collector.java

```diff
@@ -55,13 +61,34 @@ protected void acceptSpans(byte[] serializedSpans, D decoder, Callback<Void> cal
     List<S> spans;
     try {
       spans = decodeList(decoder, serializedSpans);
+      spans = filterSpans(spans, callback);
     } catch (RuntimeException e) {
       callback.onError(errorReading(e));
       return;
     }
     accept(spans, callback);
   }
 
+  /**
+   * Take the list of spans and pump them through pre-configured filters
+   *
+   * @param spans
+   */
+  List<S> filterSpans(List<S> spans, Callback<Void> callback) {
+    // These should never be processed in parallel.
+    // If you need parallel processing put that into your specific filter implementation.
+    List<S> processed = spans;
+    for (SpanFilter filter : filters) {
+      try {
+        processed = (List<S>)filter.processV1((List<Span>)processed, callback);
```

Comment:
@rmichela This casting feels dirty, but it works.

Reply:
Catching the `ClassCastException` for flow control is worse...

Maybe you can solve this problem with a [`TypeToken`](https://github.com/google/guava/wiki/ReflectionExplained#typetoken)?